module/zfs/arc.c

```diff
@@ -9092,15 +9092,16 @@ l2arc_apply_transforms(spa_t *spa, arc_buf_hdr_t *hdr, uint64_t asize,
 		 * write things before deciding to fail compression in nearly
 		 * every case.)
 		 */
-		cabd = abd_alloc_for_io(size, ismd);
-		tmp = abd_borrow_buf(cabd, size);
+		uint64_t bufsize = MAX(size, asize);
+		cabd = abd_alloc_for_io(bufsize, ismd);
+		tmp = abd_borrow_buf(cabd, bufsize);
 
 		psize = zio_compress_data(compress, to_write, &tmp, size,
 		    hdr->b_complevel);
 
 		if (psize >= asize) {
 			psize = HDR_GET_PSIZE(hdr);
-			abd_return_buf_copy(cabd, tmp, size);
+			abd_return_buf_copy(cabd, tmp, bufsize);
 			HDR_SET_COMPRESS(hdr, ZIO_COMPRESS_OFF);
 			to_write = cabd;
 			abd_copy(to_write, hdr->b_l1hdr.b_pabd, psize);
```

Comment:
Unless it compresses down to the asize level. If it fails to compress *and* the size > asize, you may be right? But also I think we could assert that would never happen? I'm not sure, though, under what circumstances `size > asize`.

Reply:
Yeah, I am thinking about the explicit case where the compression "fails" (which is this if-clause). I think there may be a way this can happen... I seem to recall hitting this case during my testing. I guess we could try an assert and see what happens with some zloop runs...

Reply:
I'll run the test. @rincebrain Can you think of a situation where this could happen and not be the result of a bug?

Reply:
Oh, no, what's happening here is that the psize of the buffer (which, until the introduction of the csize variable, was not always the same as the psize variable) is always less than or equal to the asize, because the asize is the psize rounded up to the device's ashift. I'm not sure when exactly this if block *doesn't* trigger; perhaps if the l2arc is using a different compression algorithm than the header originally did, somehow? But you can see that in this case, we use the `b_pabd` from the l1hdr, which presumably contains the original physical block from the disk, compressed to size `psize`.